### PR TITLE
Feature/seab 3321

### PR DESCRIPTION
**Pull request by garyluu**
For SEAB-3321 and SEAB-3341

This introduces an inconsistency between the ES client version and the running Elasticsearch version (7.10.x). However, there's no breaking changes it seems. 

The AWS Elasticsearch version can't be updated to a newer Elasticsearch.

This is kind of a temporary measure. Need to figure out what to do for OpenSearch vs Elasticsearch before proceeding with more drastic changes.

### CODE REVIEW

File: THIRD-PARTY-LICENSES.txt [hunk 1]
```diff
@@ -138,8 +138,9 @@ Lists of 357 third-party dependencies.
      (Apache License 2.0) Dropwizard Utility Classes (io.dropwizard:dropwizard-util:2.0.24 - http://www.dropwizard.io/2.0.24/dropwizard-bom/dropwizard-dependencies/dropwizard-parent/dropwizard-util)
      (Apache License 2.0) Dropwizard Validation Support (io.dropwizard:dropwizard-validation:2.0.24 - http://www.dropwizard.io/2.0.24/dropwizard-bom/dropwizard-dependencies/dropwizard-parent/dropwizard-validation)
      (Apache 2) EasyMock (org.easymock:easymock:4.0.1 - http://easymock.org/easymock)
-     (The Apache Software License, Version 2.0) elasticsearch-core (org.elasticsearch:elasticsearch-core:7.10.2 - https://github.com/elastic/elasticsearch)
-     (The Apache Software License, Version 2.0) elasticsearch-x-content (org.elasticsearch:elasticsearch-x-content:7.10.2 - https://github.com/elastic/elasticsearch)
+     (Elastic License 2.0) (Server Side Public License, v 1) elasticsearch-core (org.elasticsearch:elasticsearch-core:7.14.1 - https://github.com/elastic/elasticsearch)
+     (Elastic License 2.0) (Server Side Public License, v 1) elasticsearch-plugin-classloader (org.elasticsearch:elasticsearch-plugin-classloader:7.14.1 - https://github.com/elastic/elasticsearch)
```
Comment (denis-yuen) [inline THIRD-PARTY-LICENSES.txt:142]: are we using any plugins? maybe we can exclude this

File: THIRD-PARTY-LICENSES.txt [hunk 2]
```diff
@@ -260,7 +261,7 @@ Lists of 357 third-party dependencies.
      (MIT License) JUL to SLF4J bridge (org.slf4j:jul-to-slf4j:1.7.32 - http://www.slf4j.org)
      (Eclipse Public License 1.0) JUnit (junit:junit:4.13.1 - http://junit.org)
      (Apache License 2.0) JVM Integration for Metrics (io.dropwizard.metrics:metrics-jvm:4.1.25 - https://metrics.dropwizard.io/metrics-jvm)
-     (The Apache Software License, Version 2.0) lang-mustache (org.elasticsearch.plugin:lang-mustache-client:7.10.2 - https://github.com/elastic/elasticsearch)
+     (Elastic License 2.0) (Server Side Public License, v 1) lang-mustache (org.elasticsearch.plugin:lang-mustache-client:7.14.1 - https://github.com/elastic/elasticsearch)
```
Comment (denis-yuen) [inline THIRD-PARTY-LICENSES.txt:264]: Interesting, this doesn't seem to be an open source license 
https://opensource.org/node/1099

That said, I don't think its viral (as I think we stay open source even if we use non open-source components).

**Review by denis-yuen — COMMENTED** [review]
Need to take a quick look at how licensing works

File: THIRD-PARTY-LICENSES.txt (hunk 2 above)
Comment (garyluu) [reply THIRD-PARTY-LICENSES.txt:264]: it's the main reason why Amazon forked Elasticsearch

**Comment by denis-yuen** [comment]
This is the poison pill for AWS. 

> If you make the functionality of the Program or a modified version available to third parties as a service, you must make the Service Source Code available via network download to everyone at no charge, under the terms of this License. Making the functionality of the Program or modified version available to third parties as a service includes, without limitation, enabling third parties to interact with the functionality of the Program or modified version remotely through a computer network, offering a service the value of which entirely or primarily derives from the value of the Program or modified version, or offering a service that accomplishes for users the primary purpose of the Program or modified version.
> 
> “Service Source Code” means the Corresponding Source for the Program or the modified version, and the Corresponding Source for all programs that you use to make the Program or modified version available as a service, including, without limitation, management software, user interfaces, application program interfaces, automation software, monitoring software, backup software, storage software and hosting software, all such that a user could run an instance of the service using the Service Source Code you make available.

Is it a poison pill for us due to  https://github.com/dockstore/dockstore-deploy ?
https://www.logzilla.net/Elastic-Fork.html says 

> For example, if you are offering dashboards for monitoring, that are driven by Elasticsearch and Elasticsearch is in the background, it’s not clear that you are free of SSPL. Risk exists since you are still providing Elasticsearch as part of your proprietary “service.”

but maybe we're ok because Dockstore is driven by OpenSearch and we're only using the client libraries to talk to it.

**Comment by garyluu** [comment]
> This is the poison pill for AWS.
> 
> > If you make the functionality of the Program or a modified version available to third parties as a service, you must make the Service Source Code available via network download to everyone at no charge, under the terms of this License. Making the functionality of the Program or modified version available to third parties as a service includes, without limitation, enabling third parties to interact with the functionality of the Program or modified version remotely through a computer network, offering a service the value of which entirely or primarily derives from the value of the Program or modified version, or offering a service that accomplishes for users the primary purpose of the Program or modified version.
> > “Service Source Code” means the Corresponding Source for the Program or the modified version, and the Corresponding Source for all programs that you use to make the Program or modified version available as a service, including, without limitation, management software, user interfaces, application program interfaces, automation software, monitoring software, backup software, storage software and hosting software, all such that a user could run an instance of the service using the Service Source Code you make available.
> 
> Is it a poison pill for us due to https://github.com/dockstore/dockstore-deploy ?
> https://www.logzilla.net/Elastic-Fork.html says
> 
> > For example, if you are offering dashboards for monitoring, that are driven by Elasticsearch and Elasticsearch is in the background, it’s not clear that you are free of SSPL. Risk exists since you are still providing Elasticsearch as part of your proprietary “service.”
> 
> but maybe we're ok because Dockstore is driven by OpenSearch and we're only using the client libraries to talk to it.

https://www.elastic.co/blog/elastic-license-v2 probably explains it better

If I'm interpretting it right, I don't think Dockstore runs into those limitations

File: THIRD-PARTY-LICENSES.txt (hunk 2 above)
Comment (denis-yuen) [reply THIRD-PARTY-LICENSES.txt:263]: Yeah, I knew they changed the license for the server to not be open source. 
I just didn't know they made the client libraries not open source as well.

**Comment by denis-yuen** [comment]
> https://www.elastic.co/blog/elastic-license-v2 probably explains it better
> 
> If I'm interpretting it right, I don't think Dockstore runs into those limitations

They're one of the parties to the conflict though which is why I take their stuff (or AWS) with a grain of salt. 
So far we have 

Pros
* solves CVE

Cons
* a patch version upgrade might break us (as opposed to a minor version upgrade), in fact I'm not sure why this PR works in retrospect https://github.com/elastic/elasticsearch/issues/76479 https://github.com/elastic/elasticsearch/issues/76091 
* uncertainty about the license situation

I'm tempted to tag this and leave this in our back pocket until we make that "OpenSearch vs Elasticsearch" decision

**Comment by garyluu** [comment]
> > https://www.elastic.co/blog/elastic-license-v2 probably explains it better
> > If I'm interpretting it right, I don't think Dockstore runs into those limitations
> 
> They're one of the parties to the conflict though which is why I take their stuff (or AWS) with a grain of salt.
> So far we have
> 
> Pros
> 
> * solves CVE
> 
> Cons
> 
> * a patch version upgrade might break us (as opposed to a minor version upgrade), in fact I'm not sure why this PR works in retrospect [HighLevelClient 7.14.0+ Appears to Break Compatibility with Amazon Elasticsearch Service elastic/elasticsearch#76479](https://github.com/elastic/elasticsearch/issues/76479) [OSS distribution is not valid flavor in Java REST High Level Client v7.14.0 elastic/elasticsearch#76091](https://github.com/elastic/elasticsearch/issues/76091)
> * uncertainty about the license situation
> 
> I'm tempted to tag this and leave this in our back pocket until we make that "OpenSearch vs Elasticsearch" decision

I've only tested this locally with ES 7.10.2 and things seem to work. I don't think there's any easy way for me to test with AWS ES. The license situation seems more like it's a problem for those blatantly copying and reselling Elasticsearch as their own, so we _should_ be clear for that.

**Comment by garyluu** [comment]
Changed to use a slightly older version that avoids the validation issue (https://github.com/elastic/elasticsearch/blob/v7.13.3/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java doesn't appear to have that code)

**Comment by coverbeck** [comment]
> I've only tested this locally with ES 7.10.2 and things seem to work. I don't think there's any easy way for me to test with AWS ES. The license situation seems more like it's a problem for those blatantly copying and reselling Elasticsearch as their own, so we _should_ be clear for that.

Can you try with https://hub.docker.com/r/opensearchproject/opensearch/tags?page=1&ordering=last_updated ?

On the license, not a lawyer, but looks OK to me.

**Comment by garyluu** [comment]
> > I've only tested this locally with ES 7.10.2 and things seem to work. I don't think there's any easy way for me to test with AWS ES. The license situation seems more like it's a problem for those blatantly copying and reselling Elasticsearch as their own, so we _should_ be clear for that.
> 
> Can you try with https://hub.docker.com/r/opensearchproject/opensearch/tags?page=1&ordering=last_updated ?
> 
> On the license, not a lawyer, but looks OK to me.

This fails with 7.13.3 client and opensearch 1.0.1

**Comment by denis-yuen** [comment]
Taking a step back, am I correct with this summary?

* the three vulnerabilities we're trying to tackle are https://github.com/advisories/GHSA-3393-hvrj-w7v3 https://github.com/advisories/GHSA-62ww-4p3p-7fhj and to a lesser extent https://github.com/advisories/GHSA-hwvv-438r-mhvj (low severity and we have no sensitive info in elasticsearch anyway)
* https://github.com/advisories/GHSA-62ww-4p3p-7fhj doesn't affect us since again, we have no sensitive data in elasticsearch and we also don't have "Document and Field Level Security" enabled which seems to correspond to the "Fine-grained access control" setting in AWS 
* that leaves the first vulnerability which is a DDOS issue and is fixed in 17.13.3 https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-7.13.3.html 
* we're hoping that the client library is preventing users from constructing an arbitrary query that could lead to a DDOS i.e. if the grok fix is actually running on the server, and not in the client we would need to upgrade the elasticsearch server to at least 7.13.3 to clear this CVE
* AWS only supports up to ElasticSearch 7.10

**Comment by garyluu** [comment]
> Taking a step back, am I correct with this summary?
> 
> * the three vulnerabilities we're trying to tackle are [GHSA-3393-hvrj-w7v3](https://github.com/advisories/GHSA-3393-hvrj-w7v3) [GHSA-62ww-4p3p-7fhj](https://github.com/advisories/GHSA-62ww-4p3p-7fhj) and to a lesser extent [GHSA-hwvv-438r-mhvj](https://github.com/advisories/GHSA-hwvv-438r-mhvj) (low severity and we have no sensitive info in elasticsearch anyway)
> * [GHSA-62ww-4p3p-7fhj](https://github.com/advisories/GHSA-62ww-4p3p-7fhj) doesn't affect us since again, we have no sensitive data in elasticsearch and we also don't have "Document and Field Level Security" enabled which seems to correspond to the "Fine-grained access control" setting in AWS
> * that leaves the first vulnerability which is a DDOS issue and is fixed in 17.13.3 https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-7.13.3.html
> * we're hoping that the client library is preventing users from constructing an arbitrary query that could lead to a DDOS i.e. if the grok fix is actually running on the server, and not in the client we would need to upgrade the elasticsearch server to at least 7.13.3 to clear this CVE
> * AWS only supports up to ElasticSearch 7.10

Not aware of the first point, the issue this PR is trying to solve is linked way above (which is the DoS)
Yes, AWS Elasticsearch Service supports up to Elasticsearch 7.10 and Opensearch 1.0

**Review by coverbeck — APPROVED** [review]
Until/if we switch from AWS ES, should we update the docker-compose to use OpenSearch instead of ES?

**Comment by denis-yuen** [comment]
> Until/if we switch from AWS ES, should we update the docker-compose to use OpenSearch instead of ES?

I don't know.  AWS forked at 7.11, so I guess the question is if they also fixed/ported the ES fixes from ES 7.13.3.

https://github.com/opensearch-project/OpenSearch/issues/1078

Seems to indicate that we would need at least OpenSearch 1.1.0 based on the label?

**Comment by garyluu** [comment]
> > Until/if we switch from AWS ES, should we update the docker-compose to use OpenSearch instead of ES?
> 
> I don't know. AWS forked at 7.11, so I guess the question is if they also fixed/ported the ES fixes from ES 7.13.3.
> 
> [opensearch-project/OpenSearch#1078](https://github.com/opensearch-project/OpenSearch/issues/1078)
> 
> Seems to indicate that we would need at least OpenSearch 1.1.0 based on the label?

It looks like 1.x branch has the fix

**Comment by denis-yuen** [comment]
Looks like 1.0.1 based on labels for  https://github.com/opensearch-project/OpenSearch/pull/1087

**Comment by garyluu** [comment]
So I think the summary is that this PR fixes the Snyk issue but probably doesn't fix the actual issue (real world). To fix the real issue, AWS needs to update OpenSearch or we switch to a newer Elasticsearch via many different means

**Review by NatalieEO — APPROVED** [review]

**Review by david4096 — APPROVED** [review]

**Review by wshands — APPROVED** [review]

**Comment by garyluu** [comment]
Talking to Denis and David, I think we've decided to put this PR on hold (since it doesn't fix the real issue)

**Pull request closed without merging**